File: c++/insertion-sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *insertionSortList(ListNode *head) {
        /* https://oj.leetcode.com/problems/insertion-sort-list/
        Using insertion sort
        */
        
        if (head == NULL || head->next == NULL) return head;
        
        ListNode *dummy = new ListNode(0);
        dummy->next = head;
        
        while (head->next != NULL) {
            ListNode *p = dummy;
            while (p != head && p->next->val <= head->next->val) {
                p = p->next;
            }
            if (p == head) {
                head = head->next;
                continue;
            }
            
            ListNode *tmp = head->next->next;
            head->next->next = p->next;
            p->next = head->next;
            head->next = tmp;
        }
        
        head = dummy->next;
        delete dummy;
        return head;
    }
};

File: python/insertion-sort-list.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        if not head or not head.next:
            return head

        dummy = ListNode(-1)
        dummy.next = head
        cur = head.next
        while cur:
            print cur.val
            p = dummy
            while p.next and p.next.val < cur.val:
                p = p.next

            t = cur.next
            cur.next = p.next
            p.next = cur
            cur = t

        head = dummy.next
        del dummy
        return head


l = ListNode(1)
l.next = ListNode(1)
s = Solution()
r = s.insertionSortList(l)
#while r:
#    print r.val
#    r = r.next
